coderd/workspaceusage/tracker.go

```diff
@@ -0,0 +1,181 @@
+}
+
+type Option func(*Tracker)
+
+// WithLogger sets the logger to be used by Tracker.
+func WithLogger(log slog.Logger) Option {
+	return func(h *Tracker) {
+		h.log = log
+	}
+}
+
+// WithFlushInterval allows configuring the flush interval of Tracker.
+func WithFlushInterval(d time.Duration) Option {
+	return func(h *Tracker) {
+		ticker := time.NewTicker(d)
+		h.tickCh = ticker.C
+		h.stopTick = ticker.Stop
+	}
+}
+
+// WithFlushChannel allows passing a channel that receives
+// the number of marked workspaces every time Tracker flushes.
+// For testing only.
+func WithFlushChannel(c chan int) Option {
+	return func(h *Tracker) {
+		h.flushCh = c
+	}
+}
+
+// WithTickChannel allows passing a channel to replace a ticker.
+// For testing only.
+func WithTickChannel(c chan time.Time) Option {
+	return func(h *Tracker) {
+		h.tickCh = c
+		h.stopTick = func() {}
+	}
+}
+
+// Add marks the workspace with the given ID as having been used recently.
+// Tracker will periodically flush this to its configured Store.
+func (wut *Tracker) Add(workspaceID uuid.UUID) {
+	wut.mut.Lock()
+	wut.m[workspaceID] = struct{}{}
+	wut.mut.Unlock()
+}
+
+// flushLocked updates last_used_at of all current workspace IDs.
+// MUST HOLD LOCK BEFORE CALLING
+func (wut *Tracker) flushLocked(now time.Time) {
+	if wut.mut.TryLock() {
+		panic("developer error: must lock before calling flush()")
+	}
+	count := len(wut.m)
+	defer func() { // only used for testing
+		if wut.flushCh != nil {
+			wut.flushCh <- count
+		}
+	}()
+	if count == 0 {
+		wut.log.Debug(context.Background(), "nothing to flush")
+		return
+	}
+	// Copy our current set of IDs
+	ids := make([]uuid.UUID, 0)
+	for k := range wut.m {
+		ids = append(ids, k)
+	}
+	// Reset our internal map
+	wut.m = make(map[uuid.UUID]struct{})
+	// For ease of testing, sort the IDs lexically
+	sort.Slice(ids, func(i, j int) bool {
+		// For some unfathomable reason, byte arrays are not comparable?
+		return strings.Compare(ids[i].String(), ids[j].String()) < 0
+	})
+	// Set a short-ish timeout for this. We don't want to hang forever.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	// nolint: gocritic // system function
+	authCtx := dbauthz.AsSystemRestricted(ctx)
+	if err := wut.s.BatchUpdateWorkspaceLastUsedAt(authCtx, database.BatchUpdateWorkspaceLastUsedAtParams{
+		LastUsedAt: now,
+		IDs:        ids,
+	}); err != nil {
+		wut.log.Error(ctx, "failed updating workspaces last_used_at", slog.F("count", count), slog.Error(err))
+		return
+	}
+	wut.log.Info(ctx, "updated workspaces last_used_at", slog.F("count", count), slog.F("now", now))
+}
+
+func (wut *Tracker) Loop() {
+	defer func() {
+		wut.log.Debug(context.Background(), "workspace usage tracker loop exited")
+	}()
+	for {
+		select {
+		case <-wut.stopCh:
+			close(wut.doneCh)
+			return
+		case now, ok := <-wut.tickCh:
+			if !ok {
+				return
+			}
+			wut.mut.Lock()
+			wut.flushLocked(now.UTC())
+			wut.mut.Unlock()
```

Comment:
This can potentially hold the lock for a long time as there is a database action here. Which would block writes. Maybe that is ok since the caller of the api endpoint is an agent, and we run the api call on a go routine.

----

Have you considered the alternative of using a channel? Just placing the `m        map[uuid.UUID]struct{}` as a variable in the `Loop` function and having a channel for incoming workspace writes. Just buffer the `workspaceBump` and taildrop overflow if that ever happens. This would just make sure writers never block.


```golang
	m := make(map[uuid.UUID]struct{})
	for {
		select {
		case wid := <-wut.workspaceBump:
			m[wid] = struct{}{}
		case <-wut.stopCh:
			close(wut.doneCh)
			return
		case now, ok := <-wut.tickCh:
			if !ok {
				return
			}
			wut.flushLocked(now.UTC(), m)
			m = make(map[uuid.UUID]struct{})
		}
	}
```

Reply:
That's nice! Although writers could still block if enough writes come in at the same time, but that should be fairly unlikely.

Reply:
I think writers should use a select to just raildrop any overflow. It shouldn't probably happen with a good buffer? 

```golang
func (wut *Tracker) Add(workspaceID uuid.UUID) {
	select {
	case wut.workspaceBump <- workspaceID:
	default:
		wut.log.Warn(context.Background(), "workspace usage tracker is full, dropping workspace", slog.F("workspaceID", workspaceID))
	}
}
```

I do see what you are saying though. If the db call takes say an excessive 5s because of conn pool contention or something, the buffer has to potentially last that full 5s. With workspaces on 1min cadences, missing one isn't ideal.

Maybe make the db call async? So

```golang
	// Set a short-ish timeout for this. We don't want to hang forever.
	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	go func() {
		defer cancel()
		// nolint: gocritic // system function
		authCtx := dbauthz.AsSystemRestricted(ctx)
		if err := wut.s.BatchUpdateWorkspaceLastUsedAt(authCtx, database.BatchUpdateWorkspaceLastUsedAtParams{
			LastUsedAt: now,
			IDs:        ids,
		}); err != nil {
			wut.log.Error(ctx, "failed updating workspaces last_used_at", slog.F("count", count), slog.Error(err))
			return
		}
		wut.log.Info(ctx, "updated workspaces last_used_at", slog.F("count", count), slog.F("now", now))
	}()
```

The timeout will ensure it is not run twice across tick intervals. At this point, the mutex is probably comparable. And this does make unit testing more tricky?

Reply:
We can actually minimize the time we spend locked if we just lock while we copy the ids to a `[]uuid.UUID`.

Reply:
Yup yup :+1: